docs/naps/4-async-slicing.md

```diff
@@ -0,0 +1,636 @@
+
+Some of these issues were caused by a previous effort around asynchronous slicing
+in an effort to keep it isolated from the core code base.
+By contrast, our approach in this project is to redesign slicing in napari to
+provide a solid foundation for asychronous slicing and related future work like
+multi-canvas and multi-scale slicing.
+
+### Goals
+
+To summarize the scope of this project, we define a few high level goals.
+Each goal has many prioritized features where P0 is a must-have, P1 is a should-have,
+and P2 is a could-have. Some of these goals may already be achieved by napari in its
+current form, but are captured here to prevent any regression caused by this work.
+
+
+#### 1. Keep responding when slicing slow data
+
+- P0. When moving a dimension slider, the slider can still be controlled so that I can navigate to the desired location.
+	- Slider can be moved when data is in the middle of loading.
+	- Slider location does not return to position of last loaded slice after it was moved to a different position.
+- P0. When the slider is dragged, only slice at some positions so that I don’t wait for unwanted intermediate slicing positions.
+	- Once slider is moved, wait before performing slicing operation, and cancel any prior pending slices (i.e. be lazy).
+	- If we can reliably infer that slicing will be fast (e.g. data is a numpy array), consider skipping this delay.
+- P0. When slicing fails, I am notified so that I can understand what went wrong.
+    - May want to limit the number of notifications (e.g. lost network connection for remote data).
+- P1. When moving a dimension slider and the slice doesn’t immediately load, I am notified that it is being generated, so that I am aware that my action is being handled.
+	- Need a visual cue that a slice is loading.
+	- Show visual cue to identify the specific layer(s) that are loading in the case where one layer loads faster than another.
+
+
+#### 2. Clean up slice state and logic in layers
+
+- P0. Encapsulate the slice input and output state for each layer type, so that I can quickly and clearly identify those.
+	- Minimize number of (nested) classes per layer-type (e.g. `ImageSlice`, `ImageSliceData`, `ImageView`, `ImageLoader`).
+- P0. Simplify the program flow of slicing, so that developing and debugging against allows for faster implementation. 
+	- Reduce the complexity of the call stack associated with slicing a layer.
+	- The implementation details for some layer/data types might be complex (e.g. multi-scale image), but the high level logic should be simple.
+- P1. Move the slice state off the layer, so that its attributes only represent the whole data.
+	- Layer may still have a function to get a slice.
+	- May need alternatives to access currently private state, though doesn't necessarily need to be in the Layer (e.g. a plugin with an ND layer, that gets interaction data from 3D visualization , needs some way to get that data back to ND).
```

Comment:
Can you clarify what you mean byt this? That a `LayerSliceResponse` should know where it lives in the world coordinates?

Reply:
Yes, I think that's right. I believe this originally came up in a discussion with @alisterburt.

The response knows where it lives in the scene (i.e. the subspace of the world that is being visualized), but slicing is a lossy operation that means the output loses the info to get back to ND world coordinate system.

I guess if the response also stores at least some parts of the request (e.g. the bounding box that defines which region which are slicing), then that might be enough because the non-visualized dimensions are the sliced dimensions (either at an exact coordinate or spanning a range like thick slices).